pkg/pgmodel/model/label_list.go

```diff
@@ -4,24 +4,72 @@
 
 package model
 
+import (
+	"fmt"
+
+	"github.com/jackc/pgtype"
+	"github.com/timescale/promscale/pkg/pgmodel/model/pgutf8str"
+)
+
 type LabelList struct {
-	Names  []string
-	Values []string
+	Names  pgutf8str.TextArray
+	Values pgutf8str.TextArray
 }
 
 func NewLabelList(size int) *LabelList {
-	return &LabelList{Names: make([]string, 0, size), Values: make([]string, 0, size)}
+	nameElements := make([]pgtype.Text, 0, size)
+	valueElements := make([]pgtype.Text, 0, size)
+	return &LabelList{
+		// We want to avoid runtime conversion of []string to pgutf8str.TextArray. The best way to do that is
+		// to use directly the pgutf8str.TextArray under the hood.
+		// The implementations done here are kept in line with what happens in
+		// https://github.com/jackc/pgtype/blob/master/text_array.go
+		Names: pgutf8str.TextArray{
+			TextArray: pgtype.TextArray{
+				Elements:   nameElements,
+				Dimensions: []pgtype.ArrayDimension{{Length: int32(size), LowerBound: 1}},
+				Status:     pgtype.Present,
+			},
+		},
+		Values: pgutf8str.TextArray{
+			TextArray: pgtype.TextArray{
+				Elements:   valueElements,
+				Dimensions: []pgtype.ArrayDimension{{Length: int32(size), LowerBound: 1}},
+				Status:     pgtype.Present,
+			},
+		},
+	}
 }
-func (ls *LabelList) Add(name string, value string) {
-	ls.Names = append(ls.Names, name)
-	ls.Values = append(ls.Values, value)
+
+func (ls *LabelList) Add(name string, value string) error {
+	var (
+		nameT  pgutf8str.Text
+		valueT pgutf8str.Text
+	)
+	if err := nameT.Set(name); err != nil {
+		return fmt.Errorf("setting pgtype.Text: %w", err)
+	}
+	if err := valueT.Set(value); err != nil {
+		return fmt.Errorf("setting pgtype.Text: %w", err)
+	}
+	ls.Names.Elements = append(ls.Names.Elements, nameT.Text)
+	ls.Values.Elements = append(ls.Values.Elements, valueT.Text)
+	return nil
+}
+
+func (ls *LabelList) UpdateArrayDimensions() {
```

Comment:
I don't think this should be a public function. It's too easy to forget to call it. Instead, I would make LabelList.Names and LabelList.Values  private (names and values) and have a Get() function that returns (names pgutf8str.TextArray, values pgutf8str.TextArray). That function would call updateArrayDimensions() before returning. That way the user of LabelList does not have to worry about this.

Reply:
I think lengths should be set before using the type. Implementing length in Get() will mean, doing it post-execution, which might harm any dependency on total size in TextArray.